### PR TITLE
- Stop sending "Script error." - Adding try/catch blocks to public api

### PR DESCRIPTION
**Pull request by MaxShehovtsov**
Because of CORS some users are getting "Script error." with no details (url/stack trace/file name). This is not actionable. In this fix "Script error." 's are filtered out.
For our own exceptions (coming from the script on CDN - ai.0.js) a tracing has been added to report meaningful errors when they happen. Otherwise we have the same "Script error." problem.

### CODE REVIEW

**Comment by benaadams** [comment]
As aside:

https://azure.microsoft.com/en-gb/documentation/articles/app-insights-web-failures-exceptions/

> In a web browser; if you include script files from CDNs or other domains, ensure your script tag has the attribute crossorigin="anonymous" and the server sends CORS headers to get a stack trace and detail for unhandled javascript execptions from these resources.

File: JavaScript/JavaScriptSDK/Sender.ts [hunk 1]
```diff
@@ -78,42 +78,45 @@ module Microsoft.ApplicationInsights {
...
+                }
 
-            // ensure a sender was constructed
-            if (!this._sender) {
-                _InternalLogging.warn("No sender could be constructed for this environment, payload will be added to buffer." + Serializer.serialize(envelope));
-                return;
-            }
+                // ensure a sender was constructed
+                if (!this._sender) {
+                    _InternalLogging.warn("No sender could be constructed for this environment, payload will be added to buffer." + Serializer.serialize(envelope));
```
Comment (scsouthw) [inline JavaScript/JavaScriptSDK/Sender.ts]: This warning looks outdated, the payload doesn't get added to the buffer in this case.

File: JavaScript/JavaScriptSDK/Util.ts [hunk 2]
```diff
@@ -80,6 +80,18 @@
         }
 
         /**
+        * Checks if error has no meaningful data inside. Ususally such errors are received by window.onerror when error
+        * happens in a script from other domain (cross origin, CORS).
+        */
+        public static isMeaninglessError(message: string, url: string, lineNumber: number, columnNumber: number, error: Error): boolean {
+            return message == "Script error."
```
Comment (AlexBulankou) [inline JavaScript/JavaScriptSDK/Util.ts]: Is "Script error" not browser specific? Also what about localization? Is there any way we can check for it without checking for text?

File: JavaScript/JavaScriptSDK/Util.ts [hunk 3]
```diff
@@ -80,6 +80,18 @@
         }
 
         /**
+        * Checks if error has no meaningful data inside. Ususally such errors are received by window.onerror when error
+        * happens in a script from other domain (cross origin, CORS).
+        */
+        public static isMeaninglessError(message: string, url: string, lineNumber: number, columnNumber: number, error: Error): boolean {
```
Comment (AlexBulankou) [inline JavaScript/JavaScriptSDK/Util.ts]: Perhaps a better name for this API would be isCrossOriginError

Comment (scsouthw) [inline JavaScript/JavaScriptSDK/Util.ts]: I think this should be configurable. Although the browser strips most information out of this, the fact that a script on the site is throwing errors could still be useful information.

**Comment by AlexBulankou** [comment]
:shipit:

**Comment by MaxShehovtsov** [comment]
@benaadams Thanks Ben! I was actually considering crossorigin attribute too - we probably need to apply it to our snippet. The only reason I decided to go with try/catch is that there are already customers who receive lots of "Script error." - fixing it directly in script will apply the change without any snippet modification.

File: JavaScript/JavaScriptSDK/Sender.ts (hunk 1 above)
Comment (MaxShehovtsov) [inline JavaScript/JavaScriptSDK/Sender.ts]: Good catch!

File: JavaScript/JavaScriptSDK/Util.ts (hunk 2 above)
Comment (MaxShehovtsov) [inline JavaScript/JavaScriptSDK/Util.ts]: Based on telemetry it's the same text for different counties and doesn't have localized versions. Let's have it as is until we learn something else.

File: JavaScript/JavaScriptSDK/Util.ts (hunk 3 above)
Comment (MaxShehovtsov) [inline JavaScript/JavaScriptSDK/Util.ts]: @scsouthw agree - will add it in the next iteration

**Comment by benaadams** [comment]
I do get some useful errors from "Script Error"; get custom properties, get browser versions etc Some I get partial stack traces (using the crossorigin ); get number of users having it, so know where to look and how important it is to address.

So:

capture

Is mostly caused by 1 user with a dodgy toolbar installed which is interacting badly with our app.

capture2

So would probably want it as a configurable property as there is useful information there.

**Comment by MaxShehovtsov** [comment]
@benaadams Right Ben, that's useful because you have stack, file name, line number.

**Comment by benaadams** [comment]
@MaxShehovtsov ah, gochta missed the extra conditions :+1:

**Pull request merged**